### PR TITLE
ATR-401: practice dsl resolver

### PR DESCRIPTION
**Pull request by rasanjalee-dissanayaka**

### CODE REVIEW

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/HeartbeatPracticeApplication.kt [hunk 1]
```diff
@@ -1,27 +1,19 @@
...
-import com.heartbeat.practice.heartbeatpractice.model.RequestType
+import com.heartbeat.practice.heartbeatpractice.graphql.GraphQLProvider
 import com.heartbeat.practice.heartbeatpractice.service.PracticeService
 import com.heartbeathealth.support.logging
+import com.heartbeathealth.support.service.DeliveryMessage
+import com.rabbitmq.client.Delivery
 import kotlinx.coroutines.runBlocking
 import org.springframework.beans.factory.annotation.Autowired
-import org.springframework.boot.SpringBootConfiguration
 import org.springframework.boot.autoconfigure.EnableAutoConfiguration
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/HeartbeatPracticeApplication.kt:12]: This configuration is redundant, and remove @EnableAutoConfiguration at line 29

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/HeartbeatPracticeApplication.kt [hunk 2]
```diff
@@ -40,34 +32,19 @@ open class HeartbeatPracticeApplication {
     @Autowired lateinit var practiceService: PracticeService
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/HeartbeatPracticeApplication.kt:32]: These two variable never used, and can be removed.

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/HeartbeatPracticeApplication.kt [hunk 3]
```diff
@@ -40,34 +32,19 @@ open class HeartbeatPracticeApplication {
...
-                }
-                else -> throw java.lang.IllegalStateException("Execution of request $input failed.")
-            }
+            log.info("received input : $input")
+
+            val request: Request? = objectMapper?.readValue(input, Request::class.java)
+            val query = request!!.detail.text
+            log.info("query : $query")
+            val message = DeliveryMessage(Delivery(null, null,query.toByteArray()))
+            var response = message.let { graphQLProvider.executeQuery(it)}
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/HeartbeatPracticeApplication.kt:46]: This should val.

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/Utils/Utils.kt [hunk 4]
```diff
@@ -1,5 +1,6 @@
 package com.heartbeat.practice.heartbeatpractice.Utils
 
+import com.google.gson.Gson
 import com.heartbeathealth.notification.shim.NotificationService
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/Utils/Utils.kt:4]: Few unused imports can be removed.

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/Utils/Request.kt [hunk 5]
```diff
@@ -1,8 +1,8 @@
 package com.heartbeat.practice.heartbeatpractice.Utils
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/Utils/Request.kt:1]: Package name should not start with capital.

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/Utils/Utils.kt [hunk 6]
```diff
@@ -29,4 +27,19 @@ fun List<String>?.toUUDList() : List<UUID> {
 
 fun String.toUUID() : UUID {
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/Utils/Utils.kt:28]: This and toUUIDList never used, and can be removed.

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/Utils/Utils.kt [hunk 7]
```diff
@@ -29,4 +27,19 @@ fun List<String>?.toUUDList() : List<UUID> {
...
     return UUID.fromString(this)
+}
+
+fun <T> T.serializeToMap(): Map<String, Any> {
+    return convert()
+}
+
+//convert a map to a data class
+inline fun <reified T> Map<String, Any>.toDataClass(): T {
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/Utils/Utils.kt:37]: This function never used.

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt [hunk 8]
```diff
@@ -35,19 +38,24 @@ class PracticeService : BaseService() {
...
+    @Autowired
+    lateinit var userInfoService: UserInfoService
+    @Autowired
+    lateinit var mapService: GoogleMapService
+    @Autowired
+    lateinit var searchService: SearchService
+    @Autowired
+    lateinit var eventService: EventService
+    @Autowired
+    lateinit var notificationService: NotificationService
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt:50]: Never used, can be removed.

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt [hunk 9]
```diff
@@ -78,18 +87,23 @@ class PracticeService : BaseService() {
...
         val tags = listOf(PATIENT_TAG_COHORT_VCS, PATIENT_TAG_COHORT_OHS)
-        log.info("Adding supported user tags to Heartbeat Health practice", mapOf("practiceId" to heartbeatPracticeId, "tags" to tags))
+        log.info(
+            "Adding supported user tags to Heartbeat Health practice",
+            mapOf("practiceId" to heartbeatPracticeId, "tags" to tags)
+        )
         addSupportedUserTags(EMPTY_UUID, heartbeatPracticeId, tags)
     }
 
     fun isHeartbeatPractice(practiceId: UUID) =
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt:97]: remove this function

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt [hunk 10]
```diff
@@ -190,120 +219,135 @@ class PracticeService : BaseService() {
...
+        validateRoles(roles)
+        return if (userId == providerId)
+            selfAffiliateToPractice(providerId, practiceId, roles, newAffiliationId)
+        else
+            affiliateToPracticeAsAdmin(userId, providerId, practiceId, roles, newAffiliationId)
+    }
+
     private fun ProviderPracticeAffiliationEvent.isAdministrator() = roles.contains(PracticeRole.ADMINISTRATOR)
 
     private suspend fun withPracticeAdminIds(practiceId: UUID, closure: suspend (List<UUID>) -> Unit) {
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt:301]: remove

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt [hunk 11]
```diff
@@ -313,217 +357,282 @@ class PracticeService : BaseService() {
...
-//            }
-//        }
-//
-//    suspend fun getPatientPracticeIds(context: SessionContextInterface, patientId: UUID): List<UUID> =
-//            withPatientRegistrations(patientId) { registrations ->
-//                registrations.map { it.value.practiceId }
-//            }
-
-    suspend fun haveCommonPractices(providerId: UUID, patientId: UUID) = getCommonPracticeIds(providerId, patientId).isNotEmpty()
+    suspend fun getPracticeAffiliationsSync(
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt:470]: remove

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt [hunk 12]
```diff
@@ -313,217 +357,282 @@ class PracticeService : BaseService() {
...
+                                    it,
+                                    value.isVerified
+                                )
+                            }
+                        }
+                }
+            }
+        }
+
+    suspend fun getPatientPracticeIds(patientId: UUID): List<UUID> =
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt:503]: remove

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt [hunk 13]
```diff
@@ -313,217 +357,282 @@ class PracticeService : BaseService() {
...
+        withPatientRegistrations(patientId) { registrations ->
+            registrations.map { it.value.practiceId }
+        }
+
+    suspend fun haveCommonPractices(providerId: UUID, patientId: UUID) =
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt:508]: remove

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt [hunk 14]
```diff
@@ -313,217 +357,282 @@ class PracticeService : BaseService() {
...
+                withPatientRegistrations(patientId) { registrationEvents ->
+                    val providerPracticeIds =
+                        affiliationEvents.filter { it.value.isVerified }.map { it.value.practiceId }
+                    val patientPracticeIds = registrationEvents.map { it.value.practiceId }
+                    providerPracticeIds.intersect(patientPracticeIds).toList()
                 }
             }
+        }
 
     suspend fun getCommonProviderPracticeIds(providerId1: UUID, providerId2: UUID): List<UUID> =
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt:523]: remove

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt [hunk 15]
```diff
@@ -313,217 +357,282 @@ class PracticeService : BaseService() {
...
+        patientPanel(search = search, first = first, after = after) {
+            fields("id")
+        }
+    }.let { result ->
+        result.data?.patientPanel?.mapNodes {
+            PatientRegistration(it.id)
+        } ?: emptyConnection()
+    }
+
+    suspend fun isProviderAffiliatedToPractice(providerId: UUID, practiceId: UUID): Boolean =
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt:576]: remove

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt [hunk 16]
```diff
@@ -313,217 +357,282 @@ class PracticeService : BaseService() {
...
+        withProviderPracticeAffiliation(providerId, practiceId) { affiliation ->
+            (affiliation?.value?.isVerified ?: false)
+        }
+
+    suspend fun isProviderAffiliatedToPracticeAsAnyOf(
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt:581]: remove

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt [hunk 17]
```diff
@@ -313,217 +357,282 @@ class PracticeService : BaseService() {
...
+        withProviderPracticeAffiliation(providerId, practiceId) { affiliation ->
+            Triple(
+                affiliation != null,
+                affiliation?.value?.isVerified ?: false,
+                affiliation?.value?.roles?.intersect(roles)?.isNotEmpty() ?: false
+            )
+        }
 
 
     suspend fun getAffiliatedProviderIds(practiceId: UUID, onlyVerified: Boolean = true, onlyRole: PracticeRole?) =
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt:595]: remove

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt [hunk 18]
```diff
@@ -313,217 +357,282 @@ class PracticeService : BaseService() {
...
+                Event(
+                    id = UUID.randomUUID(),
+                    time = now.toOffsetDateTime(),
+                    objectId = practiceId,
+                    key = EventType.practice_supported_user_tag.toKey(),
+                    value = UserTagEvent(editingUserId, it)
+                )
+            }
         if (events.isNotEmpty()) eventService.addEvents(events)
     }
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt:638]: remove removeSupportedUserTags function

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt [hunk 19]
```diff
@@ -536,81 +645,124 @@ class PracticeService : BaseService() {
...
+        }
 
-    private suspend fun <T> withRegistrations(practiceId: UUID, closure: suspend (List<EventValue<PatientPracticeRegistrationEvent>>) -> T): T =
-            log.suspendRunWithContext(mapOf("practiceId" to practiceId)) {
-                val registrations = safeExecute("fetching practice patient registrations") {
-                    eventService.findByValueAtPathEquals(EventType.patient_practice_registration.toKey(), "{practiceId}", practiceId)
-                }?.toEventValueList<PatientPracticeRegistrationEvent>() ?: emptyList()
-                closure(registrations)
-            }
+    private suspend fun <T> withRegistrations(
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/service/PracticeService.kt:663]: remove

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/graphql/PracticeDSLResolver.kt [hunk 20]
```diff
@@ -0,0 +1,551 @@
...
+import com.heartbeat.practice.heartbeatpractice.exceptions.PracticeUpdateStatusCode
+import com.heartbeat.practice.heartbeatpractice.model.*
+import com.heartbeat.practice.heartbeatpractice.service.PracticeService
+import com.heartbeathealth.search.shim.*
+import com.heartbeathealth.support.graphql.UUIDScalar
+import com.heartbeathealth.support.graphql.gqldesc
+import com.heartbeathealth.support.graphql.resolver.*
+import com.heartbeathealth.support.logging
+import com.heartbeathealth.support.model.OpaqueConnection
+import com.heartbeathealth.support.model.emptyConnection
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/graphql/PracticeDSLResolver.kt:13]: remove this import

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/graphql/PracticeDSLResolver.kt [hunk 21]
```diff
@@ -0,0 +1,551 @@
...
+import com.heartbeathealth.support.model.toValidKeys
+import com.heartbeathealth.support.util.PHONE_WITH_EXTRA_DIGITS_PATTERN
+import com.heartbeathealth.support.util.US_STATE_PATTERN
+import com.heartbeathealth.support.util.UUID_PATTERN
+import com.heartbeathealth.support.util.convertFromMap
+import graphql.scalars.ExtendedScalars
+import kotlinx.coroutines.future.asCompletableFuture
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/graphql/PracticeDSLResolver.kt:20]: unused import

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/graphql/PracticeDSLResolver.kt [hunk 22]
```diff
@@ -0,0 +1,551 @@
...
+                val practices = practiceService.findPracticesById(listOf(practiceId))
+                log.info("practices matching: $practices")
+                practices.firstOrNull()
+            }
+
+            fetcher("practiceLocations") {
+                val inSearch = argOrNull<Map<String, Any>>("search")
+                val search = inSearch?.convertFromMap<PracticeSearchSpec>()
+                log.info("Searching practice locations")
+                search?.let { search ->
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/graphql/PracticeDSLResolver.kt:465]: name shadowed. You can rename second search in { } with different name, searchSpec

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/model/Practice.kt [hunk 23]
```diff
@@ -5,6 +5,7 @@ import com.heartbeathealth.search.shim.MarketingPreferences
 import com.heartbeathealth.search.shim.PracticeRole
 import com.heartbeathealth.search.shim.TranslatorPreferences
 import com.heartbeathealth.support.service.PhoneNumber
+import java.time.OffsetDateTime
```
Comment (benleehealth) [inline src/main/kotlin/com/heartbeat/practice/heartbeatpractice/model/Practice.kt:8]: unused import

**Review by benleehealth — CHANGES_REQUESTED** [review]
Assuming this is not live code, looks good to me. All comments are cosmetic.

File: src/main/kotlin/com/heartbeat/practice/heartbeatpractice/model/Practice.kt (hunk 23 above)
Comment (rasanjalee-dissanayaka) [reply src/main/kotlin/com/heartbeat/practice/heartbeatpractice/model/Practice.kt]: all addressed

**Pull request merged**